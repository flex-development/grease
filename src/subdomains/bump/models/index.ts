/**
 * @file Entry Point - Models
 * @module grease/bump/models
 */

export { default as RecommendedBump } from './recommended-bump.model'
