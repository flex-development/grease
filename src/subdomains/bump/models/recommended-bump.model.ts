/**
 * @file Models - RecommendedBump
 * @module grease/bump/models/RecommendedBump
 */

import { ReleaseType } from '#src/enums'
import { ifelse, type Omit } from '@flex-development/tutils'

/**
 * Recommended version bump.
 *
 * @class
 */
class RecommendedBump {
  /**
   * Total number of breaking changes committed since last release.
   *
   * @public
   * @instance
   * @member {number} breaks
   */
  public breaks: number

  /**
   * Recommended release type.
   *
   * @see {@linkcode ReleaseType}
   *
   * @public
   * @instance
   * @member {Extract<ReleaseType, 'major' | 'minor' | 'patch'>} bump
   */
  public bump: Extract<ReleaseType, 'major' | 'minor' | 'patch'>

  /**
   * Number of commits considered.
   *
   * @public
   * @instance
   * @member {number} commits
   */
  public commits: number

  /**
   * Total number of features committed since last release.
   *
   * @public
   * @instance
   * @member {number} features
   */
  public features: number

  /**
   * Create a new version bump recommendation.
   *
   * @param {Omit<RecommendedBump, 'bump'>} data - Version bump data
   */
  constructor(data: Omit<RecommendedBump, 'bump'>) {
    this.breaks = data.breaks
    this.commits = data.commits
    this.features = data.features
    this.bump = ifelse(
      this.breaks,
      ReleaseType.MAJOR,
      ifelse(this.features, ReleaseType.MINOR, ReleaseType.PATCH)
    )
  }
}

export default RecommendedBump
