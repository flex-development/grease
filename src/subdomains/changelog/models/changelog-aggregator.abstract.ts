/**
 * @file Models - AbstractChangelogAggregator
 * @module grease/changelog/models/AbstractChangelogAggregator
 */

import type { BreakingChange, Commit, Mention, Reference } from '#src/git'
import {
  hasOwn,
  select,
  type CompareResult,
  type PropertyKey
} from '@flex-development/tutils'
import type CommitGroup from './commit-group.model'
import type CommitType from './commit-type.model'

/**
 * Abstract changelog aggregator.
 *
 * @template T - Parsed commit type
 *
 * @abstract
 * @class
 */
abstract class AbstractChangelogAggregator<T extends Commit = Commit> {
  /**
   * Parsed commits.
   *
   * @public
   * @readonly
   * @instance
   * @member {ReadonlyArray<T>} commits
   */
  public readonly commits: readonly T[]

  /**
   * An array of objects representing explicitly supported commit message types,
   * and whether they should show up in the changelog.
   *
   * @see {@linkcode CommitType}
   *
   * @protected
   * @readonly
   * @instance
   * @member {ReadonlyArray<CommitType>} types
   */
  protected readonly types: readonly CommitType[]

  /**
   * Create a new changelog aggregator.
   *
   * @param {ReadonlyArray<CommitType>} types - Commit message types
   * @param {ReadonlyArray<T>?} [commits=[]] - Parsed commits
   */
  constructor(types: readonly CommitType[], commits: readonly T[] = []) {
    this.commits = Object.freeze(commits)
    this.types = Object.freeze(select(types, t => hasOwn(this.map, t.type)))
  }

  /**
   * Get a breaking changes array.
   *
   * @see {@linkcode BreakingChange}
   *
   * @public
   * @abstract
   *
   * @return {BreakingChange[]} Breaking changes array
   */
  public abstract get breaks(): BreakingChange[]

  /**
   * Get a sorted commit groups array.
   *
   * @see {@linkcode CommitGroup}
   *
   * @public
   * @abstract
   *
   * @return {CommitGroup<T>[]} Sorted commit groups array
   */
  public abstract get groups(): CommitGroup<T>[]

  /**
   * Get a parsed commit map.
   *
   * @protected
   * @abstract
   *
   * @return {Record<string, T[]>} Parsed commit map
   */
  protected abstract get map(): Record<string, T[]>

  /**
   * Get a mentions array.
   *
   * @see {@linkcode Mention}
   *
   * @public
   * @abstract
   *
   * @return {Mention[]} Mentions array
   */
  public abstract get mentions(): Mention[]

  /**
   * Get an array containing issue and pull request references.
   *
   * @see {@linkcode Reference}
   *
   * @public
   * @abstract
   *
   * @return {Reference[]} References array
   */
  public abstract get references(): Reference[]

  /**
   * Get a commit group key.
   *
   * @protected
   * @abstract
   *
   * @param {T} commit - Parsed commit
   * @return {PropertyKey} Commit group key
   */
  protected abstract groupby(commit: T): PropertyKey

  /**
   * Breaking change sorter.
   *
   * @see {@linkcode BreakingChange}
   *
   * @protected
   * @abstract
   *
   * @param {BreakingChange} brk1 - First breaking change to compare
   * @param {BreakingChange} brk2 - Second breaking change to compare
   * @return {CompareResult} Comparison result
   */
  protected abstract sortbrk(
    brk1: BreakingChange,
    brk2: BreakingChange
  ): CompareResult

  /**
   * Commit sorter.
   *
   * @protected
   * @abstract
   *
   * @param {T} cmt1 - First commit to compare
   * @param {T} cmt2 - Second commit to compare
   * @return {CompareResult} Comparison result
   */
  protected abstract sortcmt(cmt1: T, cmt2: T): CompareResult

  /**
   * Commit group sorter.
   *
   * @see {@linkcode CommitGroup}
   *
   * @protected
   * @abstract
   *
   * @param {CommitGroup<T>} g1 - First commit group to compare
   * @param {CommitGroup<T>} g2 - Second commit group to compare
   * @return {CompareResult} Comparison result
   */
  protected abstract sortgrp(
    g1: CommitGroup<T>,
    g2: CommitGroup<T>
  ): CompareResult
}

export default AbstractChangelogAggregator
