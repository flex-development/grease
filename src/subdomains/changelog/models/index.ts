/**
 * @file Entry Point - Models
 * @module grease/changelog/models
 */

export {
  default as AbstractChangelogAggregator
} from './changelog-aggregator.abstract'
export { default as ChangelogAggregator } from './changelog-aggregator.model'
export {
  default as ChangelogEntry,
  type ChangelogEntryDTO
} from './changelog-entry.model'
export {
  default as AbstractChangelogFormatter
} from './changelog-formatter.abstract'
export { default as ChangelogFormatter } from './changelog-formatter.model'
export { default as ChangelogInfile } from './changelog-infile.model'
export {
  default as ChangelogStream,
  type ChangelogStreamDTO
} from './changelog-stream.model'
export { default as CommitGroup } from './commit-group.model'
export { default as CommitType } from './commit-type.model'
