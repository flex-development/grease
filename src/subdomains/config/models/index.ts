/**
 * @file Entry Point - Models
 * @module grease/config/models
 */

export { default as GreaseConfig } from './grease-config.model'
