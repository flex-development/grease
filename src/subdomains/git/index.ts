/**
 * @file Entry Point - git
 * @module grease/git
 */

export * from './enums'
export { default as GitModule } from './git.module'
export type * from './interfaces'
export * from './models'
export * from './options'
export * from './providers'
export type * from './types'
