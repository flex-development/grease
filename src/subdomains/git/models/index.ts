/**
 * @file Entry Point - Models
 * @module grease/git/models
 */

export { default as CommitGrammar } from './commit-grammar.model'
export { default as Commit } from './commit.model'
export { default as Repository } from './repository.model'
