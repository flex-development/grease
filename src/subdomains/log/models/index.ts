/**
 * @file Entry Point - Models
 * @module grease/log/models
 */

export { default as LogObject } from './log-object.model'
