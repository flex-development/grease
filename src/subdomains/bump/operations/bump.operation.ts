/**
 * @file Operations - BumpOperation
 * @module grease/bump/operations/BumpOperation
 */

import type { BumpFile } from '#src/bump/types'
import { IsConstructor, IsReleaseVersion } from '#src/decorators'
import { GitOptions } from '#src/git'
import { PackageManifest } from '#src/models'
import type { ReleaseVersion } from '#src/types'
import {
  cast,
  defaults,
  unique,
  type Assign,
  type Pick
} from '@flex-development/tutils'
import {
  ArrayNotEmpty,
  IsArray,
  IsBoolean,
  IsIn,
  IsString
} from 'class-validator'

/**
 * Bump operation payload transfer object.
 */
type BumpOperationDTO = Assign<
  Partial<BumpOperation>,
  Pick<BumpOperation, 'release'>
>

/**
 * Version bump operation model.
 *
 * @see {@linkcode GitOptions}
 *
 * @class
 * @extends {GitOptions}
 */
class BumpOperation extends GitOptions {
  /**
   * Bump files.
   *
   * @see {@linkcode BumpFile}
   *
   * @default [PackageManifest]
   *
   * @public
   * @instance
   * @member {ReadonlyArray<BumpFile>} files
   */
  @IsConstructor({ each: true })
  @IsArray()
  @ArrayNotEmpty()
  public files: readonly [BumpFile, ...BumpFile[]]

  /**
   * Prerelease identifier.
   *
   * @see https://docs.npmjs.com/cli/commands/npm-dist-tag
   *
   * @default 'alpha'
   *
   * @public
   * @instance
   * @member {string} preid
   */
  @IsString()
  public preid: string

  /**
   * Base to be used for prerelease identifier.
   *
   * @default 1
   *
   * @public
   * @instance
   * @member {0 | 1} prestart
   */
  @IsIn([0, 1])
  public prestart: 0 | 1

  /**
   * Release type or version.
   *
   * @public
   * @instance
   * @member {ReleaseVersion} release
   */
  @IsReleaseVersion()
  public release: ReleaseVersion

  /**
   * Write version bump to manifest.
   *
   * @default false
   *
   * @public
   * @instance
   * @member {boolean} write
   */
  @IsBoolean()
  public write: boolean

  /**
   * Create a new version bump operation.
   *
   * @see {@linkcode BumpOperationDTO}
   *
   * @param {BumpOperationDTO} payload - Operation payload
   */
  constructor(payload: BumpOperationDTO) {
    super(payload)

    const {
      files,
      preid,
      prestart,
      release,
      write
    } = defaults(payload, {
      files: <BumpOperation['files']>[PackageManifest],
      preid: 'alpha',
      prestart: <BumpOperation['prestart']>1,
      write: false
    })

    this.files = cast(Object.freeze(unique(files, F => new F(this.cwd).file)))
    this.preid = preid
    this.prestart = prestart
    this.release = release
    this.write = write
  }
}

export { BumpOperation as default, type BumpOperationDTO }
