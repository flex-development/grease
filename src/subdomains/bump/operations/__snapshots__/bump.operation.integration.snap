// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`integration:bump/operations/BumpOperation > validation > should fail validation if schema is invalid 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "isString": "preid must be a string",
    },
    "property": "preid",
    "value": null,
  },
  ValidationError {
    "children": [],
    "constraints": {
      "isIn": "prestart must be one of the following values: 0, 1",
    },
    "property": "prestart",
    "value": 3,
  },
  ValidationError {
    "children": [],
    "constraints": {
      "isReleaseVersion": "release must be a release type or semantic version",
    },
    "property": "release",
    "value": "3.1",
  },
  ValidationError {
    "children": [],
    "constraints": {
      "isBoolean": "write must be a boolean value",
    },
    "property": "write",
    "value": 1,
  },
]
`;
